vad.py

```diff
@@ -53,14 +11,118 @@ def resize_to_min_size_(*tensors, dim = -1):
+
+class PrimitiveVAD:
+	def __init__(self,
+	             device: str = 'cpu',
+	             kernel_size_smooth_silence: int = 4096,
+	             kernel_size_smooth_signal: int = 128,
+	             kernel_size_smooth_speaker: int = 4096,
+	             silence_absolute_threshold: float = 0.05,
+	             silence_relative_threshold: float = 0.2,
+	             eps: float = 1e-9,
+	             normalization_percentile: float = 0.9):
+		self.device = device
+		self.kernel_size_smooth_silence = kernel_size_smooth_silence
+		self.kernel_size_smooth_signal = kernel_size_smooth_signal
+		self.kernel_size_smooth_speaker = kernel_size_smooth_speaker
+		self.silence_absolute_threshold = silence_absolute_threshold
+		self.silence_relative_threshold = silence_relative_threshold
+		self.eps = eps
+		self.normalization_percentile = normalization_percentile
+		self.required_wrapper = torch.tensor
+		self.required_type = 'float32'
+
+	def detect(self, signal: shapes.BT, keep_intersections: bool = False) -> shapes.BT:
+		assert len(signal) <= 2
+		signal = signal.to(self.device)
+
+		padding = self.kernel_size_smooth_signal // 2
+		stride = 1
+		smoothed_for_diff = F.max_pool1d(signal.abs().unsqueeze(1), self.kernel_size_smooth_signal, stride=stride, padding=padding).squeeze(1)
+
+		padding = self.kernel_size_smooth_silence // 2
+		stride = 1
+
+		# dilation
+		smoothed_for_silence = F.max_pool1d(signal.abs().unsqueeze(1), self.kernel_size_smooth_silence, stride=stride, padding=padding).squeeze(1)
+
+		# erosion
+		smoothed_for_silence = -F.max_pool1d(-smoothed_for_silence.unsqueeze(1), self.kernel_size_smooth_silence, stride=stride, padding=padding).squeeze(1)
+
+		# primitive VAD
+		signal_max = smoothed_for_diff.kthvalue(int(self.normalization_percentile * smoothed_for_diff.shape[-1]), dim=-1, keepdim=True).values
+		silence_absolute = smoothed_for_silence < self.silence_absolute_threshold
+		silence_relative = smoothed_for_silence / (self.eps + signal_max) < self.silence_relative_threshold
+		silence = silence_absolute | silence_relative
+
+		if keep_intersections or len(signal) == 1:
+			speech = ~silence
+		else:
+			diff_flat = smoothed_for_diff[0] - smoothed_for_diff[1]
+			speaker_id_bipole = diff_flat.sign()
+
+			padding = self.kernel_size_smooth_speaker // 2
+			stride = 1
+			speaker_id_bipole = F.avg_pool1d(speaker_id_bipole.view(1, 1, -1), kernel_size=self.kernel_size_smooth_speaker, stride=stride, padding=padding).view(-1).sign()
+
+			# removing 1 sample silence at 1111-1-1-1-1 boundaries, replace by F.conv1d (patterns -101, 10-1)
+			speaker_id_bipole = torch.where((speaker_id_bipole == 0) & (F.avg_pool1d(speaker_id_bipole.abs().view(1, 1, -1), kernel_size=3, stride=1, padding=1).view(-1) == 2 / 3) & (
+						F.avg_pool1d(speaker_id_bipole.view(1, 1, -1), kernel_size=3, stride=1, padding=1).view(-1) == 0), torch.ones_like(speaker_id_bipole), speaker_id_bipole)
+
+			resize_to_min_size_(silence, speaker_id_bipole, dim=-1)
```

Comment:
`resize_to_min_size_` is dangerous. for longer use, maybe better to get rid of it